File: examples/android-chat/app/src/main/java/so/openhorizon/sapient/chat/ChatViewModel.kt
package so.openhorizon.sapient.chat

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import uniffi.sapient_ffi.GenerationOptions
import uniffi.sapient_ffi.LlmSession
import uniffi.sapient_ffi.TokenListener

data class DisplayMessage(val role: Role, val text: String) {
    enum class Role { USER, ASSISTANT }
}

sealed interface Status {
    data object Idle : Status
    /** First send downloads + loads the model — can take minutes cold. */
    data class Loading(val model: String) : Status
    data object Generating : Status
    data class Failed(val message: String) : Status
}

class ChatViewModel : ViewModel() {
    val messages = MutableStateFlow<List<DisplayMessage>>(emptyList())
    val status = MutableStateFlow<Status>(Status.Idle)
    val backendLabel = MutableStateFlow<String?>(null)

    /** Dev default per docs/MOBILE.md §5.2 — smallest model until boring. */
    val modelAlias = MutableStateFlow("smollm2-135m-q4")

    private var session: LlmSession? = null
    private var loadedAlias: String? = null
    // Per-send flag: a new send must NOT reopen an old stream's cancel gate,
    // so each turn gets its own; this points at the active turn's.
    private var activeCancel: AtomicBoolean? = null
    // Bumped by every send and by Clear — stale writes from an old stream
    // (tokens, status flips) compare against it and drop.
    private val turnEpoch = AtomicInteger(0)

    val isBusy: StateFlow<Status> get() = status

    fun send(text: String) {
        val prompt = text.trim()
        if (prompt.isEmpty() || status.value is Status.Loading || status.value is Status.Generating) return

        messages.update { it + DisplayMessage(DisplayMessage.Role.USER, prompt) +
            DisplayMessage(DisplayMessage.Role.ASSISTANT, "") }
        val replyIndex = messages.value.lastIndex

        val alias = modelAlias.value
        val needsLoad = session == null || loadedAlias != alias
        status.value = if (needsLoad) Status.Loading(alias) else Status.Generating

        val epoch = turnEpoch.incrementAndGet()
        val cancelled = AtomicBoolean(false)
        activeCancel = cancelled

        val listener = object : TokenListener {
            override fun onToken(token: String): Boolean {
                messages.update { msgs ->
                    // The transcript can be cleared (and refilled by a new
                    // turn) while a stream drains — drop stale tokens.
                    if (turnEpoch.get() != epoch || replyIndex > msgs.lastIndex) return@update msgs
                    msgs.toMutableList().also {
                        it[replyIndex] = it[replyIndex].copy(text = it[replyIndex].text + token)
                    }
                }
                return !cancelled.get() // false = cancel generation engine-side
            }
        }

        // The FFI is blocking by design — keep it on the IO dispatcher.
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val active = if (needsLoad) {
                    // Greedy decoding: deterministic, least drift for tiny dev models
                    // (docs/MOBILE.md §5.2); add temperature only with 0.5B+ models.
                    LlmSession.load(alias, GenerationOptions(maxTokens = 512u))
                        .also {
                            session = it
                            loadedAlias = alias
                            backendLabel.value = it.backendLabel()
                        }
                } else {
                    session!!
                }
                if (turnEpoch.get() == epoch) status.value = Status.Generating
                active.chatStream(prompt, listener)
                if (turnEpoch.get() == epoch) status.value = Status.Idle
            } catch (e: Exception) {
                if (turnEpoch.get() != epoch) return@launch
                messages.update { msgs ->
                    if (replyIndex > msgs.lastIndex) return@update msgs
                    msgs.toMutableList().also { it[replyIndex] = it[replyIndex].copy(text = "") }
                }
                status.value = Status.Failed(e.message ?: e.toString())
            }
        }
    }

    /** Stop mid-reply; the partial text stays in the transcript by design. */
    fun stop() {
        activeCancel?.set(true)
    }

    fun clearConversation() {
        // Stop any in-flight stream and invalidate its epoch so a late token
        // or status flip from the old turn can't touch the fresh transcript.
        turnEpoch.incrementAndGet()
        activeCancel?.set(true)
        session?.reset()
        messages.value = emptyList()
        status.value = Status.Idle
    }
}

File: examples/swift-chat/Sources/SapientChatUI/ChatViewModel.swift
import Foundation
import Sapient

/// One rendered message bubble.
public struct DisplayMessage: Identifiable, Equatable {
    public enum Role { case user, assistant }
    public let id = UUID()
    public let role: Role
    public var text: String
}

/// Thread-safe cancellation flag shared with the streaming listener.
final class Cancellation: @unchecked Sendable {
    private let lock = NSLock()
    private var flag = false
    func cancel() { lock.lock(); flag = true; lock.unlock() }
    var isCancelled: Bool { lock.lock(); defer { lock.unlock() }; return flag }
}

/// Bridges the blocking FFI token callback onto the main actor.
/// Returning `false` from `onToken` cancels generation engine-side.
final class StreamListener: TokenListener, @unchecked Sendable {
    private let cancellation: Cancellation
    private let deliver: @Sendable (String) -> Void
    init(cancellation: Cancellation, deliver: @escaping @Sendable (String) -> Void) {
        self.cancellation = cancellation
        self.deliver = deliver
    }
    func onToken(token: String) -> Bool {
        deliver(token)
        return !cancellation.isCancelled
    }
}

@MainActor
public final class ChatViewModel: ObservableObject {
    public enum Status: Equatable {
        case idle
        /// First send downloads + loads the model — can take minutes cold.
        case loading(model: String)
        case generating
        case failed(String)
    }

    @Published public private(set) var messages: [DisplayMessage] = []
    @Published public private(set) var status: Status = .idle
    @Published public private(set) var backendLabel: String?
    /// Dev default per docs/MOBILE.md §5.2: iterate on the smallest model;
    /// switch to e.g. `llama3.2-1b-q4` only once the plumbing is boring.
    @Published public var modelAlias = "smollm2-135m-q4"

    private var session: LlmSession?
    private var loadedAlias: String?
    private var cancellation: Cancellation?
    /// The FFI is blocking by design — keep it off the main thread and off
    /// the Swift cooperative pool. Serial: one turn at a time.
    private let inferenceQueue = DispatchQueue(label: "so.openhorizon.sapient.inference")

    public init() {
        // Keep model downloads inside the app sandbox so the OS can reclaim
        // them and uninstall removes them (docs/MOBILE.md §5.4).
        if let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first {
            setenv("HF_HOME", caches.appendingPathComponent("sapient").path, 1)
        }
    }

    public var isBusy: Bool {
        if case .idle = status { return false }
        if case .failed = status { return false }
        return true
    }

    public func send(_ text: String) {
        let prompt = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !prompt.isEmpty, !isBusy else { return }

        messages.append(DisplayMessage(role: .user, text: prompt))
        messages.append(DisplayMessage(role: .assistant, text: ""))
        let replyIndex = messages.count - 1

        let alias = modelAlias
        let needsLoad = session == nil || loadedAlias != alias
        status = needsLoad ? .loading(model: alias) : .generating

        let cancellation = Cancellation()
        self.cancellation = cancellation
        let listener = StreamListener(cancellation: cancellation) { [weak self] token in
            DispatchQueue.main.async { self?.messages[replyIndex].text += token }
        }

        // Capture the resolved session ON the main actor — the queue closure
        // must not read isolated state (older Swift compilers reject it, and
        // Swift 6 makes it a hard error).
        let existingSession: LlmSession? = needsLoad ? nil : session

        inferenceQueue.async { [weak self] in
            do {
                let active: LlmSession
                if let existing = existingSession {
                    active = existing
                } else {
                    active = try LlmSession.load(
                        model: alias,
                        options: GenerationOptions(maxTokens: 512, temperature: 0.7)
                    )
                    DispatchQueue.main.async {
                        self?.session = active
                        self?.loadedAlias = alias
                        self?.backendLabel = active.backendLabel()
                    }
                }
                DispatchQueue.main.async { self?.status = .generating }
                _ = try active.chatStream(userMessage: prompt, listener: listener)
                DispatchQueue.main.async { self?.status = .idle }
            } catch {
                DispatchQueue.main.async {
                    self?.messages[replyIndex].text = ""
                    self?.status = .failed("\(error)")
                }
            }
        }
    }

    /// Ask the engine to stop mid-reply; the partial text stays in the
    /// transcript (and in the session history — intentional, see sapient-ffi).
    public func stop() {
        cancellation?.cancel()
    }

    public func clearConversation() {
        session?.reset()
        messages.removeAll()
        status = .idle
    }
}
